Added quote and certify functionality

Kakemone opened the pull request.

Kakemone committed:
Implements certify and quote functionality

Kakemone commented:
I think your branch wasn't completely up to date so some of these commits aren't my ones.

ionut-arm commented:
> I think your branch wasn't completely up to date so some of these commits aren't my ones.

I've tried to update and my branch now but for some reason the PR doesn't pick up the change 🤔

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,187 @@
+        nonce: Vec<u8>,
+    ) -> Result<TpmStatement> {
+        let (public_area, _, _) = context.read_public(object)?;
+        let (session_1, _, _) = context.sessions();
```
I think you should create `session_1` in this function too, instead of pulling it from `context`, because we don't know if the context we get as input has any sessions

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,187 @@
+        self.x509_chain.append(&mut certificates);
+    }
+
+    pub fn convert_to_tpmstmtformat_and_encode(&self) -> Vec<u8> {
```
calling this just `encode` should probably be good enough, we can add documentation afterwards to explain what we're encoding to

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,187 @@
+        let pub_area_value: Vec<u8>;
+        match self.signature.clone().marshall() {
+            Ok(sig) => sig_value = sig,
+            Err(_) => panic!("tried to convert Err type"),
```
Hmm, I think it'd be better to return `Result<Vec<u8>>` from this function - panicking/unwrapping is something we try to avoid at all costs because it essentially crashes the user app. Then you can just `?` out all the `Result`s you get along the way

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,187 @@
+            Ok(cert_info) => cert_info_value = cert_info,
+            Err(_) => panic!("tried to convert Err type"),
+        };
+        let mut alg_value = -1;
+        if let SignatureScheme::RsaSsa { hash_scheme: _ } = self.algorithm {
+            alg_value = -65535
+        };
```
Could you pull this part out into a function that takes in `SignatureScheme` and spits out the appropriate int value? We can then document the function to explain where the hardcoded values are coming from

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,187 @@
+            "pubArea" => pub_area_value,
+            "certInfo" => cert_info_value,
+        })
+        .unwrap()
```
Will need to figure out what error code to return from here instead of unwrapping.

ionut-arm replied:
Perhaps [this](https://github.com/parallaxsecond/rust-tss-esapi/blob/main/tss-esapi/src/error/wrapper.rs#L27), and an error log that converting the value to CBOR failed?

ionut-arm commented on tss-esapi/src/abstraction/transient/key_attestation.rs:
```diff
@@ -188,6 +188,24 @@ impl TransientKeyContext {
     }
+
+    pub fn quote(&mut self, key: ObjectWrapper, nonce: Vec<u8>) -> Result<web_authn::TpmPlatStmt> {
+        let key_handle = self.load_key(key.params, key.material, key.auth)?;
```
You'll need to flush out this handle after creating the `TpmPlatStmt` (whether it succeeded or not)

ionut-arm replied:
And same for the keys in `certify`!

ionut-arm reviewed (commented):
Thanks for the changes! 👍🏻

Kakemone committed:
Made requested changes

Changes made:
 - Create `session_1` in the `quote()` and `certify()` functions rather than getting it from `context`
 - Change the name of `convert_to_tpmstmtformat_and_encode()` and `convert_to_tpmplatstmtformat_and_encode()` to `encode()`
 - Change the output type of `quote()` and `certify()  from `Vec<u8>` to `Result<Vec<u8>>`
 - Pull the fetching of `alg` into a separate function

Kakemone committed:
Flushed remaining unflushed handles

Kakemone committed:
Remove `unwrap()` call

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -67,41 +75,34 @@ impl TpmStatement {
+        if let SignatureScheme::RsaSsa { hash_scheme: _ } = algorithm {
+            alg_value = -65535
+        };
+        alg_value
```
you can make this more Rusty:
```
match algorithm {
    SignatureScheme::RsaSsa { .. } => -65535,
    _ => -1,
}
```

Kakemone committed:
Fixed flushing

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

Kakemone committed:
Removed unnecessary and erroneous session

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

Kakemone committed:
Fix conversion of cbor value into Vec<u8>

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

Kakemone committed:
Floated selection_list to parsec
Floated selection_list so that it is generated within the parsec service from the values in the config file

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,181 @@
+            SymmetricDefinition::AES_128_CFB,
+            HashingAlgorithm::Sha256,
+        )?;
+        let signing_scheme = SignatureScheme::RsaSsa {
+            hash_scheme: HashScheme::new(HashingAlgorithm::Sha256),
+        };
```
We're gonna need to make this dependent on the parameters of `key`, but not necessarily now. 

If you want to have a go at this, you can try to read the [public area](https://docs.rs/tss-esapi/7.1.0/tss_esapi/structures/enum.Public.html) of `key` (as you did for `object`), then convert the [`RsaScheme`](https://docs.rs/tss-esapi/7.1.0/tss_esapi/structures/struct.PublicRsaParameters.html#method.rsa_scheme) or [`EccScheme`](https://docs.rs/tss-esapi/7.1.0/tss_esapi/structures/struct.PublicEccParameters.html#method.ecc_scheme) (from the `parameters` field in the public area) to [`SignatureScheme`](https://docs.rs/tss-esapi/7.1.0/tss_esapi/structures/enum.SignatureScheme.html). You have to do this manually, there's no `From`/`TryFrom`...

Otherwise, just add a TODO comment

ionut-arm replied:
(and same applies for the `TpmPlatStatement` constructor too)

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,181 @@
+        let pub_area = self.public_area.clone().marshall()?;
+        let cert_info = self.attestation.clone().marshall()?;
+        let alg = TpmStatement::get_alg_value(self.algorithm);
+        let mut encoded_certificate = vec![];
```
`encoded_token` would probably be a better name here, "certificate" normally means the x509 type.

ionut-arm reviewed (commented):
Thanks, looks good! Left a few minor comments, but overall looks good, looking forward to some tests/docs :D

Kakemone committed:
Makes 'signing_scheme' dependent on the parameters of 'key' and some other small fixes

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

Kakemone committed:
Fixes 'get_sig_scheme'

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

Kakemone committed:
Moves 'get_sig_scheme' out of 'TpmStatement' and 'TpmPlatStmt'

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,192 @@
+        )?;
+        let (key_public_area, _, _) = context.read_public(key)?;
+        let signing_scheme = get_sig_scheme(key_public_area)?;
+        let (attestation, signature) = context
+            .execute_with_sessions((session_1, None, None), |ctx| {
```
We don't really need to name it `session_1` given that there's only one, and we can just use `execute_with_session` instead here.

ionut-arm commented on tss-esapi/src/abstraction/web_authn.rs:
```diff
@@ -0,0 +1,192 @@
+        }
+    }
+
+    fn get_alg_value(algorithm: SignatureScheme) -> i32 {
+        match algorithm {
+            SignatureScheme::RsaSsa { .. } => -65535,
+            _ => -1,
+        }
+    }
```
Can probably do with this the same as with `get_sig_scheme` and just take it out as an independent function.

ionut-arm reviewed (approved):
Left a couple of comments (just to keep track of them somewhere), but I'll merge this now and we can address the comments in a future, clean-up PR
